### DISCUSSION TITLE
status of module objects

### DISCUSSION THREAD

**Discussion by MoonShiesty** [aptos-labs/aptos-developer-discussions#56] [Questions] [answered]
i noticed `aptos move create-object-and-publish-package` is available in the latest aptos cli with documentation in [aptos-core](https://github.com/aptos-labs/aptos-core/blob/93ed9aefff550250ecc620651cd3eb4cf37ecc40/aptos-move/framework/aptos-framework/doc/object_code_deployment.md?plain=1#L4). this is a feature im super excited about

what is the status of module objects (can i use in testnet, mainnet)?
are there any examples of how to retrieve a signer_cap for a module object?

**Comment by 0x-j** [comment]
Module object will be shipped as part of Aptos v1.10. It's already live on devnet, and we plan to release it on 03/28 on mainnet.

**Comment by gregnazario** [comment]
Objects use a different approach from resource accounts.

Resource accounts can only retrieve the signer from the signer capability, where the module objects use a reference based on ownership of the code object. 

@JohnChangUK someone's interested in your work and you can give more details :)

**Comment by JohnChangUK** [comment, marked as answer]
As @0xaptosj  said, in devnet already and plans to release to mainnet along with the governance proposal execution, as this feature is gated. Aptos CLI v2.5 will also need to be released as this contains the commands to execute this deployment successfully.

You cannot retrieve the `signer_cap` as objects don't use this. In the `object_code_deployment` module, the `ManagingRefs` keeps reference od the `ExtendRef`, which is used to generate the object signer. 

Upgrades and freezing (making the code immutable) is done via object ownership, i.e. you must be the owner of the code. The retrieving signers and auth mechanism is abstracted away from the user, you simply just call `upgrade` if you own the code :)

**Reply by MoonShiesty** [reply]
awesome, the upgrade abstractions look really nice.

we also use the `signer_cap` to do things like withdraw and manage module owned fa stores, transfer DAs and other objects it creates. with `resource_account` modules we retrieve and store the cap in  `init_module`: 
```
fun init_module(account: &signer) {
    let signer_cap = resource_account::retrieve_resource_account_cap(account, @protocol_deployer);
    move_to(account, PermissionConfig {
        signer_cap,
    });
}
```

will the framework provide similar support for module objects to store and later use their signer_caps?

**Reply by JohnChangUK** [reply]
I see, we can plan to expose the `signer` for allowing developers to extend the module object, rather than just having code in the object. Is there a reason you want to extend the module object, i.e. have more metadata in the code object? Else you could create another object for that :)

**Reply by MoonShiesty** [reply]
i want a signer_cap so the module can manage objects it owns. for example: protocol FA stores, transferring NFTs with the module as the creator, see this discussion for an example: https://github.com/aptos-labs/aptos-developer-discussions/discussions/54 

or aptos's package manager example: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/move-examples/package_manager/sources/package_manager.move.

**Reply by JohnChangUK** [reply]
I see, the recommended way of doing this is to have objects have ownership/control over other objects. We want developers to start using this pattern rather than the `SignerCapability`, as the latter is more error prone. We refer to this term as `composable`, as objects own other objects. 

Here is an example of how to create an object, which owns collections. The module you define can have access to mint/transfer NFTs and other things: https://github.com/aptos-labs/daily-move/blob/main/snippets/composable-nfts/sources/composable_nfts.move#L85-L92

**Reply by MoonShiesty** [reply]
i want to be able to [create a collection](https://github.com/aptos-labs/daily-move/blob/main/snippets/composable-nfts/sources/composable_nfts.move#L293) with @deploy_addr (as the module object) as the signer/creator/owner. we could create a second object (ObjectController) but right now we don't need to do that for resource_accounts.

id prefer to see the module directly own objects, rather creating a second object to manage ownership. can you expand on why having the module directly own objects is more prone to errors?

**Reply by JohnChangUK** [reply]
The problem with resource accounts is that if you forget to store the `SignerCapability`, then there is no way to generate the `signer`, and this has caused problems in the past. 

And ah, I see what you are saying, you don't want to create another object to manage ownership of objects. We've discussed about having the Object code deployed to an object having exposure to the `ExtendRef/signer` for the owner to extend this object. This way, you can get the object's `signer` and have this be the owner of objects. Hope that answers your question :)

**Reply by MoonShiesty** [reply]
> We've discussed about having the Object code deployed to an object having exposure to the ExtendRef/signer for the owner to extend this object. This way, you can get the object's signer and have this be the owner of objects. 

+1 on this. this would be our only blocker to picking up module objects
